### PR TITLE
Use smart rar_basename() instead of list comprehensions

### PR DESCRIPTION
**Pull request by medariox**
- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read the [contribution guide](https://github.com/pymedusa/Medusa/blob/master/.github/CONTRIBUTING.md)

### CODE REVIEW

**Comment by fernandog** [comment]
As i said `process_media(path, videoInRar, nzbName, process_method, force, is_priority, ignore_subs, result)`  passed videoInRar, and `process_media` calls list_associated_files. so you don't need to call RAR again to check rar contents

File: medusa/post_processor.py [hunk 1]
```diff
@@ -287,6 +288,16 @@ def list_associated_files(self, file_path, base_name_only=False, subtitles_only=
 
         return file_path_list
 
+    @staticmethod
+    def rar_basename(filepath, filelist):
+        videofile = os.path.basename(filepath)
+        rars = (x for x in filelist if rarfile.is_rarfile(x))
+
+        for rar in rars:
+            content = rarfile.RarFile(rar).namelist()
```
Comment (fernandog) [inline medusa/post_processor.py]: As i said `process_media(path, videoInRar, nzbName, process_method, force, is_priority, ignore_subs, result)`  passed videoInRar, and `process_media` calls list_associated_files. so you don't need to call RAR again to check rar contents

Comment (medariox) [inline medusa/post_processor.py]: We don't want to get the name of the video in the rar, but the name of the rar that holds the video we are processing. That's exactly what the code does.

Comment (fernandog) [inline medusa/post_processor.py]: The name of the rar is in file_list.
I used that in my code.
again the import rar is useless

Comment (fernandog) [inline medusa/post_processor.py]: rar_file = [os.path.basename(f).rpartition('.')[0].lower() for f in file_list if helpers.get_extension(f).lower() == 'rar']

Comment (medariox) [inline medusa/post_processor.py]: Already explained: we can't and don't need to assume that the rar file in the folder is the one that contains the video file we are processing. We can make sure it is with this code. It has nearly no impact on resources and avoids incorrect associations.

Comment (fernandog) [inline medusa/post_processor.py]: Already told you. Will explain one more time.
The function is called with videoInRar. Which is the video extracted from RAR.

File: medusa/post_processor.py [hunk 2]
```diff
@@ -277,26 +226,77 @@ def list_associated_files(self, file_path, base_name_only=False, subtitles_only=
...
                 self._delete(extensions_to_delete)
         else:
-            self._log(u'No associated files for {0} were found during this pass'.format(file_path), logger.DEBUG)
+            self._log(u'No associated files for {0} were found during this pass'.format(filepath), logger.DEBUG)
 
         return file_path_list
 
     @staticmethod
-    def rar_basename(filepath, filelist):
+    def _search_files(path, pattern='*', subfolders=None, base_name_only=None, sort=False):
```
Comment (labrys) [inline medusa/post_processor.py]: Shouldn't we just add pathlib as a dependency and use `Path.glob` here?

Comment (medariox) [inline medusa/post_processor.py]: It's not possible, because it's not recursive in py2.

Comment (labrys) [inline medusa/post_processor.py]:
```Starting with Python 3.4, one can use the glob() method of one of the Path classes in the new pathlib module, which supports ** wildcards. For example:

from pathlib import Path

for file_path in Path('src').glob('**/*.c'):
    print(file_path) # do whatever you need with these files

**Comment by medariox** [comment]
@labrys 
As requested I've added pathlib2 which requires scandir.

**Comment by fernandog** [comment]
@medariox didn't worked:

Error:
```
Packed releases detected: [u'quantico.211.720p-dimension.rar']
Unpacking archive: quantico.211.720p-dimension.rar
Failed Unrar archive quantico.211.720p-dimension.rar: exception: access violation writing 0x00003EA0
```

Manual PP logs:
https://gist.github.com/fernandog/4f78a75c77791b8d2aeff495927cdf33

Logs:
```
2017-02-21 17:02:12 DEBUG    Thread-23 :: [51ff22a] Destination folder for this episode: C:\Users\Fernando\Documents\Séries\Quantico\Season 02
2017-02-21 17:02:12 DEBUG    Thread-23 :: [51ff22a] Checking if the path C:\Users\Fernando\Documents\Séries\Quantico\Season 02 already exists
2017-02-21 17:02:12 DEBUG    Thread-23 :: [51ff22a] Folder C:\Users\Fernando\Documents\Séries\Quantico\Season 02 didn't exist, creating it
2017-02-21 17:02:12 DEBUG    Thread-23 :: [51ff22a] Found the following associated files for C:\Users\Fernando\Documents\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\Quantico.S02E11.720p.HDTV.X264-DIMENSION.mkv: [u'C:\\Users\\Fernando\\Documents\\PP\\Quantico.S02E11.720p.HDTV.X264-DIMENSION\\Quantico.S02E11.720p.HDTV.X264-DIMENSION.pt-BR.srt']
2017-02-21 17:02:12 DEBUG    Thread-23 :: [51ff22a] Hard linking file from C:\Users\Fernando\Documents\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\Quantico.S02E11.720p.HDTV.X264-DIMENSION.mkv to C:\Users\Fernando\Documents\Séries\Quantico\Season 02\Quantico.S02.E11.ZRTORCH.720p.HDTV.x264.mkv
2017-02-21 17:02:12 DEBUG    Thread-23 :: [51ff22a] Moving file from C:\Users\Fernando\Documents\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\Quantico.S02E11.720p.HDTV.X264-DIMENSION.pt-BR.srt to C:\Users\Fernando\Documents\Séries\Quantico\Season 02\Quantico.S02.E11.ZRTORCH.720p.HDTV.x264.pt-BR.srt 
```

**Comment by medariox** [comment]
@fernandog
Need debug logs. The code doesn't try to unpack the rars, so this shouldn't happen at all. Maybe because of new rarfile version? Changed the exception in the new version or something.

**Comment by fernandog** [comment]
Full logs:
https://gist.github.com/fernandog/f8b049c86206b0cbdb810abfbb0d4d1a

**Comment by medariox** [comment]
@fernandog 
From the log you can see that the error is not caused by the new code, but by some old code in process_tv. I'm not going to touch process_tv, as that code is a complete mess and the time required to make it at least decent would be more than I'm currently able and willing to spend on it. That's why I reverted rarfile to version 2.8, please let me know if you still encounter issues.

**Comment by fernandog** [comment]
But in same scenario, files and etc it works with develop. I tested those exact same files with my PR

**Comment by medariox** [comment]
Like I said, it was caused by the updated rarfile and not by the new code.

**Comment by fernandog** [comment]
Tomorrow will test again

File: medusa/post_processor.py [hunk 3]
```diff
@@ -362,17 +364,14 @@ def _combined_file_operation(self, file_path, new_path, new_base_name, associate
...
         # base name with file path (without extension and ending dot)
-        old_base_name = file_path.rpartition('.')[0]
+        old_base_name = os.path.splitext(file_path)[0]
         old_base_name_length = len(old_base_name)
 
         for cur_file_path in file_list:
             # remember if the extension changed
             changed_extension = None
             # file extension without leading dot (for example: de.srt)
             extension = cur_file_path[old_base_name_length + 1:]
```
Comment (fernandog) [inline medusa/post_processor.py]: extension here for the nfo = u''

so when hardlinking the file, its moving without extension:

Quantico\Season 02\Quantico.S02.E11.ZRTORCH.720p.HDTV.x264. 

```
2017-02-22 09:44:42 DEBUG    Thread-27 :: [51ff22a] Deleting file C:\Users\Fernando\Documents\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\quantico.211.720p-dimension.sfv
2017-02-22 09:44:42 DEBUG    Thread-27 :: [51ff22a] Moving file from C:\Users\Fernando\Documents\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\Quantico.S02E11.720p.HDTV.X264-DIMENSION.mkv to C:\Users\Fernando\Documents\Séries\Quantico\Season 02\Quantico.S02.E11.ZRTORCH.720p.HDTV.x264.mkv 
2017-02-22 09:44:42 DEBUG    Thread-27 :: [51ff22a] Moving file from C:\Users\Fernando\Documents\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\quantico.211.720p-dimension.nfo to C:\Users\Fernando\Documents\Séries\Quantico\Season 02\Quantico.S02.E11.ZRTORCH.720p.HDTV.x264. 
2017-02-22 09:44:42 DEBUG    Thread-27 :: [51ff22a] Moving file from C:\Users\Fernando\Documents\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\Quantico.S02E11.720p.HDTV.X264-DIMENSION.pt-BR.srt to C:\Users\Fernando\Documents\Séries\Quantico\Season 02\Quantico.S02.E11.ZRTORCH.720p.HDTV.x264.pt-BR.srt 
```

image

image

**Comment by fernandog** [comment]
why did you removed the valid language subtitle check?

` only add subtitles with valid languages to the list`

ping @ratoaq2

**Comment by medariox** [comment]
The check is useless as we now basically support all kind of codes and it wasn't working anyway.

**Comment by fernandog** [comment]
Looks good now:
```
2017-02-22 14:43:17 DEBUG    [51ff22a] Deleting file ~\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\quantico.211.720p-dimension.sfv
2017-02-22 14:43:17 DEBUG    [51ff22a] Deleting file ~\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\Quantico.S02E11.720p.HDTV.X264-DIMENSION.pt-BR.mkv
2017-02-22 14:44:29 DEBUG    [51ff22a] Moving file from ~\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\Quantico.S02E11.720p.HDTV.X264-DIMENSION.mkv to ~\Séries\Quantico\Season 02\Quantico.S02.E11.ZRTORCH.720p.HDTV.x264.mkv 
2017-02-22 14:44:31 DEBUG    [51ff22a] Moving file from ~\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\quantico.211.720p-dimension.nfo to ~\Séries\Quantico\Season 02\Quantico.S02.E11.ZRTORCH.720p.HDTV.x264.nfo 
2017-02-22 14:44:39 DEBUG    [51ff22a] Moving file from ~\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\Quantico.S02E11.720p.HDTV.X264-DIMENSION.en.srt to ~\Séries\Quantico\Season 02\Quantico.S02.E11.ZRTORCH.720p.HDTV.x264.en.srt 
2017-02-22 14:44:57 DEBUG    [51ff22a] Moving file from ~\PP\Quantico.S02E11.720p.HDTV.X264-DIMENSION\Quantico.S02E11.720p.HDTV.X264-DIMENSION.pt-BR.srt to ~\Séries\Quantico\Season 02\Quantico.S02.E11.ZRTORCH.720p.HDTV.x264.pt-BR.srt 
2017-02-22 14:44:57 DEBUG    [51ff22a] 295515: Setter sets location to ~\Séries\Quantico\Season 02\Quantico.S02.E11.ZRTORCH.720p.HDTV.x264.mkv
```

**Pull request merged**